File: go/datastructures/binary_search_tree.go
package datastructures

import "cmp"

// BinarySearchTree implements a binary search tree.
// Time: O(log n) average, O(n) worst for all operations
// Space: O(n)
type BinarySearchTree[T cmp.Ordered] struct {
	root *bstNode[T]
	size int
}

type bstNode[T any] struct {
	value T
	left  *bstNode[T]
	right *bstNode[T]
}

func NewBinarySearchTree[T cmp.Ordered]() *BinarySearchTree[T] {
	return &BinarySearchTree[T]{}
}

func (t *BinarySearchTree[T]) Insert(value T) {
	t.root = t.insertNode(t.root, value)
	t.size++
}

func (t *BinarySearchTree[T]) insertNode(node *bstNode[T], value T) *bstNode[T] {
	if node == nil {
		return &bstNode[T]{value: value}
	}

	if value < node.value {
		node.left = t.insertNode(node.left, value)
	} else if value > node.value {
		node.right = t.insertNode(node.right, value)
	}
	return node
}

func (t *BinarySearchTree[T]) Contains(value T) bool {
	return t.findNode(t.root, value) != nil
}

func (t *BinarySearchTree[T]) findNode(node *bstNode[T], value T) *bstNode[T] {
	if node == nil {
		return nil
	}

	if value < node.value {
		return t.findNode(node.left, value)
	} else if value > node.value {
		return t.findNode(node.right, value)
	}
	return node
}

func (t *BinarySearchTree[T]) Remove(value T) bool {
	sizeBefore := t.size
	t.root = t.removeNode(t.root, value)
	return t.size < sizeBefore
}

func (t *BinarySearchTree[T]) removeNode(node *bstNode[T], value T) *bstNode[T] {
	if node == nil {
		return nil
	}

	if value < node.value {
		node.left = t.removeNode(node.left, value)
	} else if value > node.value {
		node.right = t.removeNode(node.right, value)
	} else {
		t.size--

		if node.left == nil {
			return node.right
		}
		if node.right == nil {
			return node.left
		}

		minRight := t.findMinNode(node.right)
		node.value = minRight.value
		node.right = t.removeNode(node.right, minRight.value)
		t.size++ // Compensate for recursive remove
	}
	return node
}

func (t *BinarySearchTree[T]) findMinNode(node *bstNode[T]) *bstNode[T] {
	for node.left != nil {
		node = node.left
	}
	return node
}

func (t *BinarySearchTree[T]) findMaxNode(node *bstNode[T]) *bstNode[T] {
	for node.right != nil {
		node = node.right
	}
	return node
}

func (t *BinarySearchTree[T]) Min() (T, bool) {
	var zero T
	if t.root == nil {
		return zero, false
	}
	return t.findMinNode(t.root).value, true
}

func (t *BinarySearchTree[T]) Max() (T, bool) {
	var zero T
	if t.root == nil {
		return zero, false
	}
	return t.findMaxNode(t.root).value, true
}

func (t *BinarySearchTree[T]) InOrder() []T {
	result := make([]T, 0, t.size)
	t.inOrderTraverse(t.root, &result)
	return result
}

func (t *BinarySearchTree[T]) inOrderTraverse(node *bstNode[T], result *[]T) {
	if node == nil {
		return
	}
	t.inOrderTraverse(node.left, result)
	*result = append(*result, node.value)
	t.inOrderTraverse(node.right, result)
}

func (t *BinarySearchTree[T]) PreOrder() []T {
	result := make([]T, 0, t.size)
	t.preOrderTraverse(t.root, &result)
	return result
}

func (t *BinarySearchTree[T]) preOrderTraverse(node *bstNode[T], result *[]T) {
	if node == nil {
		return
	}
	*result = append(*result, node.value)
	t.preOrderTraverse(node.left, result)
	t.preOrderTraverse(node.right, result)
}

func (t *BinarySearchTree[T]) PostOrder() []T {
	result := make([]T, 0, t.size)
	t.postOrderTraverse(t.root, &result)
	return result
}

func (t *BinarySearchTree[T]) postOrderTraverse(node *bstNode[T], result *[]T) {
	if node == nil {
		return
	}
	t.postOrderTraverse(node.left, result)
	t.postOrderTraverse(node.right, result)
	*result = append(*result, node.value)
}

func (t *BinarySearchTree[T]) IsEmpty() bool {
	return t.size == 0
}

func (t *BinarySearchTree[T]) Size() int {
	return t.size
}

func (t *BinarySearchTree[T]) Clear() {
	t.root = nil
	t.size = 0
}

File: kotlin/src/main/kotlin/datastructures/BinarySearchTree.kt
package datastructures

/**
 * Binary Search Tree implementation.
 * Time: O(log n) average, O(n) worst for all operations
 * Space: O(n)
 */
class BinarySearchTree<T : Comparable<T>>(
    private val comparator: Comparator<T>? = null
) {
    private class Node<T>(var value: T, var left: Node<T>? = null, var right: Node<T>? = null)

    private var root: Node<T>? = null
    private var count = 0

    fun insert(value: T) {
        root = insertNode(root, value)
        count++
    }

    private fun insertNode(node: Node<T>?, value: T): Node<T> {
        if (node == null) return Node(value)

        val cmp = compare(value, node.value)
        when {
            cmp < 0 -> node.left = insertNode(node.left, value)
            cmp > 0 -> node.right = insertNode(node.right, value)
        }
        return node
    }

    fun contains(value: T): Boolean = findNode(root, value) != null

    private fun findNode(node: Node<T>?, value: T): Node<T>? {
        if (node == null) return null

        val cmp = compare(value, node.value)
        return when {
            cmp < 0 -> findNode(node.left, value)
            cmp > 0 -> findNode(node.right, value)
            else -> node
        }
    }

    fun remove(value: T): Boolean {
        val sizeBefore = count
        root = removeNode(root, value)
        return count < sizeBefore
    }

    private fun removeNode(node: Node<T>?, value: T): Node<T>? {
        if (node == null) return null

        val cmp = compare(value, node.value)
        when {
            cmp < 0 -> node.left = removeNode(node.left, value)
            cmp > 0 -> node.right = removeNode(node.right, value)
            else -> {
                count--
                when {
                    node.left == null -> return node.right
                    node.right == null -> return node.left
                    else -> {
                        val minRight = findMinNode(node.right!!)
                        node.value = minRight.value
                        node.right = removeNode(node.right, minRight.value)
                        count++ // Compensate for the recursive call decrement
                    }
                }
            }
        }
        return node
    }

    private fun findMinNode(node: Node<T>): Node<T> {
        var current = node
        while (current.left != null) {
            current = current.left!!
        }
        return current
    }

    private fun findMaxNode(node: Node<T>): Node<T> {
        var current = node
        while (current.right != null) {
            current = current.right!!
        }
        return current
    }

    fun min(): T? = root?.let { findMinNode(it).value }

    fun max(): T? = root?.let { findMaxNode(it).value }

    fun inOrder(): List<T> {
        val result = mutableListOf<T>()
        inOrderTraverse(root, result)
        return result
    }

    private fun inOrderTraverse(node: Node<T>?, result: MutableList<T>) {
        if (node == null) return
        inOrderTraverse(node.left, result)
        result.add(node.value)
        inOrderTraverse(node.right, result)
    }

    fun preOrder(): List<T> {
        val result = mutableListOf<T>()
        preOrderTraverse(root, result)
        return result
    }

    private fun preOrderTraverse(node: Node<T>?, result: MutableList<T>) {
        if (node == null) return
        result.add(node.value)
        preOrderTraverse(node.left, result)
        preOrderTraverse(node.right, result)
    }

    fun postOrder(): List<T> {
        val result = mutableListOf<T>()
        postOrderTraverse(root, result)
        return result
    }

    private fun postOrderTraverse(node: Node<T>?, result: MutableList<T>) {
        if (node == null) return
        postOrderTraverse(node.left, result)
        postOrderTraverse(node.right, result)
        result.add(node.value)
    }

    fun isEmpty(): Boolean = count == 0

    fun size(): Int = count

    fun clear() {
        root = null
        count = 0
    }

    private fun compare(a: T, b: T): Int = comparator?.compare(a, b) ?: a.compareTo(b)
}
